src/Domain/FunctionsOfTime/FunctionOfTimeHelpers.cpp

```diff
@@ -45,15 +46,18 @@ void StoredInfo<MaxDerivPlusOne, false>::pup(PUP::er& p) {
   p | stored_quantities;
 }
 
-void reset_expiration_time(const gsl::not_null<double*> prev_expiration_time,
-                           const double next_expiration_time) {
-  if (next_expiration_time < *prev_expiration_time) {
+void reset_expiration_time(
+    const gsl::not_null<std::atomic<double>*> prev_expiration_time,
+    const double next_expiration_time) {
+  const double loaded_prev_expr_time =
```

Comment:
This all isn't safe because the atomic can get change after the `if` check. You want a compare and swap: https://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange

Reply:
It's safe because we don't allow concurrent writing.  A lot of things will break if we allow that.

Reply:
Given the number of issues we've had with these sorts of assumptions, I don't think it's worth the risk of not making the code threadsafe by switching to a compare and swap.

Reply:
Fixing it here is reasonable, but the other case is in a function that is intentionally not threadsafe for performance reasons.

Reply:
Compare and swap mostly is an issue when scaling in a highly contended environment. If we were updating the FoTs ~1 per RHS evaluation per element it _might_ matter, but given the number of bugs this file alone has produced, I still think it's better to do the check and actually have some way of knowing we hit a threading issue than not.

I also don't remember discussing making the function not threadsafe for performance reasons, I thought it was because of the underlying physics/design of the control system as a whole. I remember not wanting to use locks because of performance reasons, but that's separate for a CAS on write.

Reply:
You're right, I was conflating the two things.  We could lock the function to make it threadsafe without a significant performance penalty.

Reply:
So I don't think this function specifically is actually used at all. From a quick search, it's only actually used in tests.

Reply:
Two separate things:

Should we just remove this function? If not, we can keep it and just do the CAS since this function sort of acts like that's what it wants to do. I don't have a preference since I don't know enough about why this was added in the first place.

----

@wthrowe a few thoughts on this:
1. What do you think about adding a `SPECTRE_DEBUG_THREADS` macro that we can enable to enable various threading-related checks? We might even want to have a dox page/section listing some of the possible checks to do inside this macro guard that help with detecting threading issues.
2. Assuming 1. (or using `SPECTRE_DEBUG`) we could add a lock only in the threading-debug context, then do a `try_lock` on it and `ERROR` if the lock can't be obtained. This would tell us we are somehow violating an assumption we've baked into the code. I worry we (I?) will break this subtly with nodegroup elements by accident, so having a sanity check could be great.

Reply:
Let's just do it unconditionally.  Managing a lock with no contention has to be a lot cheaper than the multiple memory allocations and several O(total number of updates)-cost operations already done by that function.

Annoyingly, `try_lock` is allowed to just randomly return false because it feels like it.  I imagine that in practice it only happens when multiple threads are manipulating the lock in close succession, but we don't have a guarantee.  If we read the current expiration time at the start before locking and then do a CAS at the end I think that catches everything we can.

(Don't care if we keep this function.)

Reply:
I think I'll remove this function because I don't see a use for it. Presumably this was added very early on when we didn't know how to deal with expiration times, so being able to change only the expiration time (and not the derivative) seemed like a good idea at the time. However, now we know (more or less) how everything is going to work, and it doesn't seem like we'll need this function.

Now for the `update` function, I'll get the current expiration time at the beginning of the function and then do a CAS at the end. This seems like the best we can do? Anything else I should add?

Reply:
Let's make it thread-safe.  It's, I think, three lines and maybe a comment, and maybe we'll remove the lock in the cache someday to support something that can handle simultaneous writes.

So do the CAS stuff discussed above and then just stick a lock_guard at the start of the function.

Reply:
That sounds great, thank you!